### PR TITLE
Add feature update existing urls to db adapter

### PR DESCRIPTION
**Pull request by oatovar**
* Add UpdateURL method signature to URL interface

* Implement UpdateURL in db adapter

* Add integration tests to db adapter

### Description
This allows the db adapter to update the URL table which is necessary for the url mutate feature.

### CODE REVIEW

**Comment by byliuyang** [comment]
@oatovar Please fix code climate's comment

**Comment by oatovar** [comment]
URLFake does not implement the UpdateURL method which is causing the build to fail. I will be working on adding this.

File: backend/app/usecase/repository/url.go [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
+	"time"
+
+	"github.com/short-d/short/app/entity"
+)
 
 // URL accesses urls from storage, such as database.
 type URL interface {
 	IsAliasExist(alias string) (bool, error)
 	GetByAlias(alias string) (entity.URL, error)
+	// TODO change to CreateURL
```
Comment (byliuyang) [inline backend/app/usecase/repository/url.go:13]: Please find our TODO format here: [s/swe-onboarding](http://s/swe-onboarding)

Comment (oatovar) [reply backend/app/usecase/repository/url.go:13]: Will create the issue and update the TODO to match short format.

File: backend/app/adapter/db/url.go [hunk 2]
```diff
@@ -39,6 +40,7 @@ WHERE "%s"=$1;`,
 }
 
 // Create inserts a new URL into url table.
+// TODO change to CreateURL
```
Comment (byliuyang) [inline backend/app/adapter/db/url.go:43]: Let's put TODO on top of the function comment.

File: backend/app/adapter/db/url.go [hunk 3]
```diff
@@ -61,6 +63,44 @@ VALUES ($1, $2, $3, $4, $5);`,
 	return err
 }
 
+// UpdateURL updates a URL that exists within the URL table and returns the newly updated URL if there is no error found while
+// committing the update.
+func (u *URLSql) UpdateURL(key string, newAlias string, newOriginalURL string, expireAt *time.Time) (entity.URL, error) {
```
Comment (byliuyang) [inline backend/app/adapter/db/url.go:68]:
```suggestion
func (u *URLSql) UpdateURL(oldAlias string, newURL entity.URL) (entity.URL, error) {
```

File: backend/app/adapter/db/url.go [hunk 4]
```diff
@@ -61,6 +63,44 @@ VALUES ($1, $2, $3, $4, $5);`,
 	return err
 }
 
+// UpdateURL updates a URL that exists within the URL table and returns the newly updated URL if there is no error found while
```
Comment (byliuyang) [inline backend/app/adapter/db/url.go:66]:
```suggestion
// UpdateURL updates a URL that exists within the url table.
```
The types on the function signature are very clear.

File: backend/app/adapter/db/url.go [hunk 5]
```diff
@@ -61,6 +63,44 @@ VALUES ($1, $2, $3, $4, $5);`,
...
+SET "%s"=$1, "%s"=$2, "%s"=$3
+WHERE "%s"=$4;`,
+		table.URL.TableName,
+		table.URL.ColumnAlias,
+		table.URL.ColumnOriginalURL,
+		table.URL.ColumnExpireAt,
+		key,
+	)
+
+	row := u.db.QueryRow(
```
Comment (byliuyang) [inline backend/app/adapter/db/url.go:80]: Should this be execute?

File: backend/app/adapter/db/url.go [hunk 6]
```diff
@@ -61,6 +63,44 @@ VALUES ($1, $2, $3, $4, $5);`,
...
+	)
+
+	url := entity.URL{}
+	err := row.Scan(
+		&url.Alias,
+		&url.OriginalURL,
+		&url.ExpireAt,
+		&url.CreatedAt,
+		&url.UpdatedAt,
+	)
```
Comment (byliuyang) [inline backend/app/adapter/db/url.go:95]: You can return `newURL` directly.

File: backend/app/adapter/db/url_integration_test.go [hunk 7]
```diff
@@ -175,6 +175,7 @@ func TestURLSql_GetByAlias(t *testing.T) {
 	}
 }
 
+// TODO change to TestURLSql_CreateURL
```
Comment (byliuyang) [inline backend/app/adapter/db/url_integration_test.go:178]: Format

File: backend/app/adapter/db/url_integration_test.go [hunk 8]
```diff
@@ -242,6 +243,123 @@ func TestURLSql_Create(t *testing.T) {
 	}
 }
 
+func TestURLSql_UpdateURL(t *testing.T) {
+	createdAt := mustParseTime(t, "2017-05-01T08:02:16-07:00")
+	now := time.Now()
+
+	testCases := []struct {
+		name           string
+		key            string
```
Comment (byliuyang) [inline backend/app/adapter/db/url_integration_test.go:252]:
```suggestion
		oldAlias            string
```

File: backend/app/adapter/db/url_integration_test.go [hunk 9]
```diff
@@ -242,6 +243,123 @@ func TestURLSql_Create(t *testing.T) {
...
+		newAlias       string
+		newOriginalURL string
+		expireAt       *time.Time
```
Comment (byliuyang) [inline backend/app/adapter/db/url_integration_test.go:255]:
```suggestion
		newURL entity.URL
```

File: backend/app/adapter/db/url_integration_test.go [hunk 10]
```diff
@@ -242,6 +243,123 @@ func TestURLSql_Create(t *testing.T) {
...
+					alias:     "220uFicCJj",
+					longLink:  "https://www.google.com",
+					createdAt: &createdAt,
+				},
+			},
+			hasErr:      true,
+			expectedURL: entity.URL{},
+		},
+		{
+			name: "alias already exists",
```
Comment (byliuyang) [inline backend/app/adapter/db/url_integration_test.go:274]:
```suggestion
			name: "alias is taken",
```

File: backend/app/adapter/db/url_integration_test.go [hunk 11]
```diff
@@ -242,6 +243,123 @@ func TestURLSql_Create(t *testing.T) {
...
+					createdAt: &createdAt,
+				},
+			},
+			hasErr: false,
+			expectedURL: entity.URL{
+				Alias:       "220uFicCJj",
+				OriginalURL: "https://www.facebook.com",
+				UpdatedAt:   &now,
+			},
+		},
```
Comment (byliuyang) [inline backend/app/adapter/db/url_integration_test.go:326]: I think this one is unnecessary unless we provide an test case for invalid long link.

File: backend/app/adapter/db/userurl.go [hunk 12]
```diff
@@ -44,6 +44,7 @@ func (u UserURLRelationSQL) FindAliasesByUser(user entity.User) ([]string, error
 
 	var aliases []string
 	rows, err := u.db.Query(statement, user.Email)
+	// TODO errors should be checked before using defer
```
Comment (byliuyang) [inline backend/app/adapter/db/userurl.go:47]: Format

File: backend/app/usecase/repository/url_fake.go [hunk 13]
```diff
@@ -63,6 +64,33 @@ func (u URLFake) GetByAliases(aliases []string) ([]entity.URL, error) {
 	return urls, nil
 }
 
+// UpdateURL updates a URL that exists within the URL table and returns the newly updated URL if there is no error found while
```
Comment (byliuyang) [inline backend/app/usecase/repository/url_fake.go:67]:
```suggestion
// UpdateURL updates an existing URL with new properties.
```

File: backend/app/usecase/repository/url_fake.go [hunk 14]
```diff
@@ -63,6 +64,33 @@ func (u URLFake) GetByAliases(aliases []string) ([]entity.URL, error) {
...
+// committing the update.
```
Comment (byliuyang) [inline backend/app/usecase/repository/url_fake.go:68]:
```suggestion
```

**Comment by byliuyang** [comment]
Hi @oatovar, thank you so much for working this cool feature! I left some comments on the PR!

File: backend/app/adapter/db/url.go (hunk 2 above)
Comment (oatovar) [reply backend/app/adapter/db/url.go:43]: If I put the TODO at the top then code climate will fail.

Comment (byliuyang) [reply backend/app/adapter/db/url.go:43]: @oatovar Okey. Then please keep it like this.

File: backend/app/adapter/db/url.go (hunk 3 above)
Comment (oatovar) [reply backend/app/adapter/db/url.go:68]: Will update the interface and implementations to use the new arguments.

File: backend/app/adapter/db/url.go (hunk 5 above)
Comment (oatovar) [reply backend/app/adapter/db/url.go:80]: Yes thank you for catching that 💯

**Comment by oatovar** [comment]
@byliuyang I have completed the requested changes, please let me know if there is anywhere else I can improve.

File: backend/app/usecase/repository/url_fake.go [hunk 15]
```diff
@@ -63,6 +64,26 @@ func (u URLFake) GetByAliases(aliases []string) ([]entity.URL, error) {
 	return urls, nil
 }
 
+// UpdateURL updates an existing URL with new properties.
+func (u URLFake) UpdateURL(oldAlias string, newURL entity.URL) (entity.URL, error) {
+	prevURL, ok := u.urls[oldAlias]
+	if !ok {
+		return entity.URL{}, errors.New("URL to update not found")
```
Comment (byliuyang) [inline backend/app/usecase/repository/url_fake.go:71]:
```suggestion
		return entity.URL{}, errors.New("alias not found")
```

**Pull request merged**